[NoQA] Add tests for switch to Expensify classic flow

hungvu193 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Explanation of Change
<!-- Explain what your change does and how it addresses the linked issue -->
Add tests for recently added logic:
```javascript
    if (shouldOpenBookACall) {
        Navigation.navigate(ROUTES.SETTINGS_EXIT_SURVERY_BOOK_CALL.route);
        return;
    }
    Navigation.navigate(ROUTES.SETTINGS_EXIT_SURVEY_CONFIRM.route);
```
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/51703
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. From your root project, run the following command:
```javascript
TZ=utc NODE_OPTIONS=--experimental-vm-modules jest tests/ui/InitialSettingPageTest.tsx
```
2. Verify that all the tests passed.
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
N/A
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image

It's acceptable to write "Same as tests" if the QA team is able to run the tests in the above "Tests" section.
--->
// TODO: These must be filled out, or the issue title must include "[No QA]."
No QA
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/eac675c7-eb69-4761-9d19-80a476c7f8b6

</details>

hungvu193 committed:
add test for switch to expensify classic flow

hungvu193 committed:
rename file

hungvu193 committed:
improve test

hungvu193 commented on tests/unit/InitialSettingPageTest.tsx:
```diff
@@ -0,0 +1,116 @@
+            await act(async () => {
+                await TestHelper.signInWithTestUser(USER_A_ACCOUNT_ID, USER_A_EMAIL, undefined, undefined, 'A');
+            });
+            await act(async () => {
+                await Onyx.set(ONYXKEYS.NVP_TRYNEWDOT, {
+                    classicRedirect: {
+                        dismissed: dismissedValue,
+                    },
+                });
+            });
```
@mountiny Do you know anyone familiar with jest that can help?
Even I tried to set the `dismissed` here to true or false. When I tried to log the value of `shouldOpenBookACall` in `InitialSettingPage` while running the test, the value of `shouldOpenBookACall` was always true.

hungvu193 committed:
fix lint

mountiny commented:
@hungvu193 how is this going?

hungvu193 commented:
Still stuck on the second test, I will raise a slack convo to get some helps.

hungvu193 committed:
Merge remote-tracking branch 'origin/main' into followup/51703

hungvu193 committed:
Update the testcase

hungvu193 committed:
remove unused mock

hungvu193 commented:
Oh weird. I think you can ignore this PR @Ollyws, I'm not sure why you're assigned, @shubham1206agra is the C+ on the linked issue, he will handle this one. Ty

hungvu193 commented:
My initial tests were:
1. Render the <App />.
2. Go to Settings.
3. Trigger the `Switch To Expensify Class Button`.

However while debugging the tests, the logic inside [action](https://github.com/Expensify/App/blob/7cb72604f0129e2771ac72bd83d1b33e9b80d594/src/pages/settings/InitialSettingsPage.tsx#L247-L254) of `Switch To Expensify Class Button` was never triggered, which made me stuck there. Any suggestions?

hungvu193 commented:
> @hungvu193 Can you check if action was mocked anywhere?

I don't think I mocked `action`, this is the old test btw
https://github.com/Expensify/App/blob/1fa2c4e15881b380ea7116f5166b834904a731c9/tests/unit/InitialSettingPageTest.tsx

hungvu193 committed:
make tests running

hungvu193 commented:
Updated the tests @shubham1206agra.

hungvu193 committed:
remove unused import

hungvu193 commented:
> @hungvu193 Can you add this test to test suite so it will run in CI?

How to do that? Sorry I'm new with this one.



> @hungvu193 Is there a reason why test takes 50 sec to run?

Mine takes 4.5 seconds to run both tests. I think every test case needs to re-render and login the App (also wait for Onyx data..etc) it probably takes longer based on our connection?

<img width="583" alt="Screenshot 2024-11-22 at 16 29 17" src="https://github.com/user-attachments/assets/f37e182b-26bf-4260-b974-f278341a7413">

hungvu193 committed:
move to UI section

hungvu193 commented:
> Sorry, I got confused. Can you put this test under ui folder since test is related to UI?

Done 😄

hungvu193 committed:
Merge remote-tracking branch 'origin/main' into followup/51703

mountiny commented on tests/ui/InitialSettingPageTest.tsx:
Is this a correct file name?

mountiny reviewed (dismissed):
Code looks good to me but its not clear to me why the test is name this way?

hungvu193 commented:
> Code looks good to me but its not clear to me why the test is name this way?

Do you have any suggestions? I was about to name it "SwitchToExpensifyClassicFlow" but then I changed my mind cause I think in the future we might add more tests to this file whenever we added new logic into InitialSettingPage.

hungvu193 committed:
rename test

hungvu193 committed:
Merge branch 'followup/51703' of https://github.com/hungvu193/App into followup/51703

mountiny reviewed (approved):
Thanks!
